node/network/approval-distribution/src/lib.rs

```diff
@@ -1221,61 +1266,17 @@ impl ApprovalDistribution {
 				},
 			};
 			match message {
-				FromOverseer::Communication {
-					msg: ApprovalDistributionMessage::NetworkBridgeUpdateV1(event),
-				} => {
-					state.handle_network_msg(&mut ctx, &self.metrics, event).await;
-				},
-				FromOverseer::Communication {
-					msg: ApprovalDistributionMessage::NewBlocks(metas),
-				} => {
-					gum::debug!(target: LOG_TARGET, "Processing NewBlocks");
-					state.handle_new_blocks(&mut ctx, &self.metrics, metas).await;
-				},
-				FromOverseer::Communication {
-					msg: ApprovalDistributionMessage::DistributeAssignment(cert, candidate_index),
-				} => {
-					gum::debug!(
-						target: LOG_TARGET,
-						"Distributing our assignment on candidate (block={}, index={})",
-						cert.block_hash,
-						candidate_index,
-					);
-
-					state
-						.import_and_circulate_assignment(
-							&mut ctx,
-							&self.metrics,
-							MessageSource::Local,
-							cert,
-							candidate_index,
-						)
-						.await;
-				},
-				FromOverseer::Communication {
-					msg: ApprovalDistributionMessage::DistributeApproval(vote),
-				} => {
-					gum::debug!(
-						target: LOG_TARGET,
-						"Distributing our approval vote on candidate (block={}, index={})",
-						vote.block_hash,
-						vote.candidate_index,
-					);
-
-					state
-						.import_and_circulate_approval(
-							&mut ctx,
-							&self.metrics,
-							MessageSource::Local,
-							vote,
-						)
-						.await;
-				},
+				FromOverseer::Communication { msg } =>
+					Self::handle_incoming(&mut ctx, state, msg, self.metrics.clone()).await,
 				FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate {
+					deactivated,
 					..
 				})) => {
 					gum::trace!(target: LOG_TARGET, "active leaves signal (ignored)");
 					// handled by NewBlocks
+					for deactivated in deactivated {
+						state.recent_outdated_heads.note_outdated(deactivated);
```

Comment:
This doesn't make sense with how the approval distribution subsystem works. We care about all unfinalized blocks, unlike statement distribution where we only care about the head of the chain.

We only prune stuff on finality, so that's where `note_outdated` should be triggered.

Reply:
I'd suggest using a different mechanism in `RecentOutdatedBlocks` (and a rename, because these aren't 'heads') where we just keep the last 2 batches of block hashes that were removed on finality.

Reply:
Fixed.